// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package hw2;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service MessengerServer {

  rpc Login (Request) returns (Reply) {}
  rpc List (Request) returns (ListReply) {}
  rpc Join (Request) returns (Reply) {}
  rpc Leave (Request) returns (Reply) {}
  // Bidirectional streaming RPC
  rpc Chat (stream Message) returns (stream Message) {} 
  rpc SendCredentials (Credentials) returns (Credentials) {}
}

// Enables communication between servers
service ServerChat{
	// Asks for a reply and restarts the server if no replay is recieved
	rpc pulseCheck (Reply) returns (Reply) {}
  rpc DataSend (Reply) returns (Reply) {}
  rpc DataSendFollowers(Request) returns (Reply){}
  rpc UpdateDatabase (Request) returns (Reply) {}
  rpc dataSync (DataSync) returns (DataSync) {}
}
message ListReply {
  repeated string all_rooms = 1;
  repeated string joined_rooms = 2;
}

message Request {
  string username = 1;
  repeated string arguments = 2;
}

message Credentials {
  string hostname = 1;
  string portnumber = 2;
  string confirmation = 3;
}

message Reply {
  string msg = 1;
}

message Message {
  //unique ID
  string id = 1;
  //Username who sent the message
  string username = 2;
  //Message from the user
  string msg = 3;
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 4;
}

message DataSync {
  //unique ID
  repeated string ids = 1;
  //The message
  repeated string messages = 2;
}